Add possibility for custom git user name/email config #61

haugene opened the pull request:
### Checklist

- [x] All new jobs, commands, executors, parameters have descriptions
- [x] Examples have been added for any significant new features
- [x] README has been updated, if necessary

### Motivation, issues

Follow-up from https://github.com/CircleCI-Public/orb-tools-orb/issues/61

### Description

Extracting a command for setting git config and enabling custom git config, with fallback to todays values. Backwards compatible.

I checked the "examples and README" part of the checklist as this is a minor change. But as I propose in my issue #61 this might also be added as parameters to the jobs that use it, and then it would also show up as an option on the orb registry page.

I'll make appropriate changes to the PR if needed.

KyleTryon commented:
Hello @haugene, 

Am I correct in thinking/understand that the " **_$GIT_CONFIG_USER_NAME_**" and "**_$GIT_CONFIG_USER_EMAIL_**" Would be set as environment variables manually as a fallback? If that is so, we may want to include parameters in the git-config description which would alert the users via the Orb registry that they should create these environment variables.

https://circleci.com/docs/2.0/orb-author-faq/#environment-variables

haugene commented:
Hey @KyleTryon,

Thanks for the feedback and sorry about the delay. You're right - I was thinking it should be a configurable fallback and that it behaves as now per default. Added them as parameters now as you suggested.

Also saw that the link I provided in the issue was to a private repo build, fixed that now. In the last builds here I'm just adding and removing a simple command and changing between our version and the latest official version of the orb. Seems to work fine 👍 https://circleci.com/bb/cloudwheel/workflows/example-orb

Also found a bug where I cross-referenced the email/name parameters, but it should be good now.
Published my version of the orb here, so you can see the doc pages: https://circleci.com/orbs/registry/orb/greenbird/orb-tools

haugene commented:
Can confirm that this works in our project orbs, will close #61 

I do however want to leave a closing remark on your merge strategy. Of all the things I love about open source being able to become a contributor with any change - big or small - is a really great feeling.

Now don't get me wrong. This change was for work, and we needed it - so it's good just to avoid maintaining a fork. But getting rebased out of your change felt a bit unnecessary and it leaves the experience less than it could have been ;)

Until next time. Thanks for a great tool, this was just what we needed to get our orb development on the right track.

iynere commented:
@haugene hi, thank you, this is good feedback ! we want contributors to feel recognized, as well.

to clarify, this pattern (which is not ideal but which we've adopted in a few open-source repos) is not about rebasing folks out of their changes, but rather about working around the security issues w.r.t testing PRs from community forks. we want to run our full suite of CI tests against everyone's PRs, but the integration tests for orb projects involve publishing a dev version of the orb to our namespace & then running tests using that dev version—& obviously, community contributors do not have perms to publish orbs to our internal namespace.

so we run a limited suite of tests against forked PRs, merge those PRs to a staging branch if those tests pass, & then run our full test suite against the staging branch, before merging to master which will actually result in a new orb semantic version being published if that test suite is all green.

hope that all makes sense. we are working on a better way of handling this, with direct recognition / visibility of community contributions as one definite goal, but this is the best we've got right now. 

we super appreciate your contributions, especially to a complicated orb like orb-tools !

haugene commented:
@iynere Thanks for the clarification. I had a feeling it could be security related, and definitely not out to get peoples' commits, but just wanted to nudge the issue anyways as I thought it could be an oversight.

I'm also maintaining my own open source project and that was where I started experimenting with CircleCI. I met the same issue with wanting to test pull requests but not give that kind of playground for anyone then trying to retrieve the secrets such a test would need to have available.
I would however think that with a human "sanity check" of changes before pushing to a staging branch would be enough to enforce a level of control and still avoid rebasing - but there are probably things I'm not considering.

That being said, I'm not discouraged from contributing again - far from it - it was a good experience with good feedback and a fast merge. I've just experienced in my own project that merging peoples commits (no matter how much I need to rewrite it first 😆) makes them that much more eager to keep on contributing. But, I understand it's a work in progress and I'm excited to see where it all goes.

Big fan of your build service, the orbs and support - so thanks again for a great platform!

iynere commented:
@haugene totally, that makes sense ! one thing i think we can do to help mitigate this is just recognize contributors in a more official manner, e.g.—

https://github.com/CircleCI-Public/circleci-cli/blob/master/CONTRIBUTORS.md
